### PR TITLE
[Discuss] Add support to Rad Racer II (4screen mirroring, scanline IRQ)

### PR DESCRIPTION
**Pull request by grimoire**
Hi,
I am working on support Rad Racer II. Which include  4 screen and scanline IRQ.

before:

uXjf7ex4SI

After:

tSkLZLW0sk
E3ovBQVBFu

There are still some glitch, I guess it has something to do with the way I add scanline IRQ in PPU. But I have no idea how to fix it.

I have add extra RAM in mapper to support nametable 2 and 3. It would be much easer to just resize the picturebus. But I think we should `emulate` what the NES do, right? Would it be better to add another mirroring type `MapperControled` so we can support more complex mirroring for some Mapper?

### CODE REVIEW

**Comment by amhndu** [comment]
Thank you for the pull request! 

> I have add extra RAM in mapper to support nametable 2 and 3. It would be much easer to just resize the picturebus. But I think we should emulate what the NES do, right? Would it be better to add another mirroring type MapperControled so we can support more complex mirroring for some Mapper?

I think this approach is good, since it's mapper specific, the extra memory should go there. I like the idea of MapperControlled!

File: src/MapperMMC3.cpp [hunk 1]
```diff
@@ -272,6 +286,14 @@ void MapperMMC3::scanline()
 	}	
 }
 
+void MapperMMC3::scanlineIRQ(){
+	scanline();
+	if(irqState()){
+		m_interruptCallback(int(CPU::InterruptType::IRQ));
```
Comment (amhndu) [inline src/MapperMMC3.cpp:292]: Why cast to int here instead of passing the enum itself?

Comment (grimoire) [reply src/MapperMMC3.cpp:292]: There will be a loop include between `CPU.h` and `Mapper.h`. Can I move the enum to a `Constant.h` so both of them can use it?

Comment (amhndu) [reply src/MapperMMC3.cpp:292]: Why would there be a loop? Mapper doesn't doesn't on CPU

Comment (grimoire) [reply src/MapperMMC3.cpp]: Done.

**Comment by amhndu** [comment]
Hey, sorry for the delay. I'm currently cleaning up some mappers, including MMC3, unfortunately I have little time, so I'm not done yet. I'll try to clean that as soon as possible. I also just found out that nametable isn't correctly read from an iNES file. I'll need to fix that before this will work well.

**Comment by grimoire** [comment]
[MMC5](https://www.nesdev.org/wiki/MMC5) has a complex mirroring mode which is configured by a mapping register, that is why I add this new enum (maybe renaming the FourScreen would be better?).

I am refactoring [PPU steps](https://github.com/grimoire/SimpleNES/blob/refactor-ppu/src/PPU.cpp) following [PPU rendering](https://www.nesdev.org/wiki/PPU_rendering) because I thought it has something to do with the glitch (trigger the IRQ by A13). But the glitch still exists after I finish(almost) the refactor. Do you have any idea what should I do to fix it?

By the way, do you need a github hook? I have some experiments on that.

**Comment by amhndu** [comment]
> GitHub hooks
That sounds great! Some contributed files have bad indentation, this would really help

**Comment by grimoire** [comment]
Sure. I will add the hooks after you cleaning up the mappers.

**Comment by amhndu** [comment]
I've pushed up some changes to master, including a fix for ines header reading. (Although I've yet to find a Rad Racer dump that has the correct 4 screen header). 
And getting back to NameTableMirroring, I suggest removing MapperControlled, and using FourScreen instead. FourScreen should mean mapper controlled. That's the only it's possible and that's the nesdev wiki seems to be calling it.

**Comment by amhndu** [comment]
> But the glitch still exists after I finish(almost) the refactor. Do you have any idea what should I do to fix it?

I also see glitches in SMB2 and Kirby's Adventure, it's possible that MMC3 itself has a bug. 
Another functionality missing is the APU, it's possible that some ROMs are using that for timing. Working on the APU is also on my queue.

**Comment by SuzieQQ** [comment]
> Sure. I will add the hooks after you cleaning up the mappers.

Hi I saw that you fixed IRQ did you try rad racer 2 which uses mapper 4 you can try to see if it goes in alien 3 game should go now

**Comment by grimoire** [comment]
> you can try to see if it goes in alien 3 game should go now

Alien 3 failed with a black screen. I guess that is caused by the `MC-ACC` chip which has different IRQ behavior. I am trying to find a fix.

**Comment by grimoire** [comment]
@SuzieQQ Alien 3 works now.

**Comment by SuzieQQ** [comment]
> @SuzieQQ Alien 3 works now.
as soon as i can i try it, now i don't remember i had a small list with IRQ bugs with other games maybe super mario 2 should go in game now! this is a particular game they say

**Comment by SuzieQQ** [comment]
4577CCFD-60D0-47B7-85C7-4FA49E203A1A
-368f2f8a-f5b1-41a1-ac2f-f4ace9a7ddd7.png)
874CDCC5-E0E3-4F07-808D-D18FF3996F09
343744B6-AC06-49EC-AE2B-2CA96F3C4C82
4977835E-82D4-49F3-8955-DAF3E2C044CB
D1D5473B-45CA-4DCE-87FF-9F0BCFC2F8CF
570915BF-6F86-4DF9-92FC-356D04D3E7FF
4CB43A41-D362-4767-942F-8AA50BEBD09F
40452B97-413D-4863-95F7-65AD6BBB9E72
0359B5F5-88B7-4954-84EA-F99AB5EE2EA0

85627310-3677-4E61-B1C6-255421DBC08B

i have tested all the games i had tried in the past these are the improvements, the rest either behaves the same, or some no longer work.
I noticed that in particular all the ninja turtle games never worked

**Comment by SuzieQQ** [comment]
7BDABA39-F5AE-4978-B1E5-B7A2EB7B1D36
84E2F32E-99B6-4878-8C1F-B7FF47D1CE1E
F53C1567-27AB-476F-9890-E71A735BC60A
CD7972CE-690A-4704-926E-006D641CE75F
C42005C0-4022-489B-B908-D8E63033DE81
F3008136-AB34-4E7B-9CE9-EEC0C237DCAD
333BBC14-52D5-4D58-A6A2-0AF3864EED69
D70F6FEA-ECE7-464B-835F-B830A7D9F7CB
87D830F0-EA12-4FC5-AEEA-EF418264E8DC
EE6B5BF1-567C-4127-845B-213232F2A361
5D35BCF9-E569-4AEF-B638-68D785BAB301

i have tested all the games i had tried in the past these are the improvements, the rest either behaves the same, or some no longer work.
I noticed that in particular all the ninja turtle games never worked

**Comment by amhndu** [comment]
@SuzieQQ Does SMB3 work now?

**Comment by SuzieQQ** [comment]
> @SuzieQQ Does SMB3 work now?

no, before it went into the menu now graphical glitch and does not enter the menu

**Comment by grimoire** [comment]
@amhndu @SuzieQQ 
Hi, I think I find out why SMB3 not working.
I download the rom from [here](https://www.emulatorgames.net/roms/nintendo/super-mario-bros-3-pc10/) and the `head[6] & 0x2` is 0, that means no extended RAM.
But after IRQ, CPU will try to read $0x7964, which leads to 0 because there is no such RAM. After I force the RAM enable, the game works.
IiFPcr2dy6

EGTylpRvW7
 
I am not sure if it is caused by the ROM(maybe this is a hacked one). Any advice?

**Comment by amhndu** [comment]
Great find! I think a simpler fix could be to simply always have extended ram available. There's no particular reason to "disable" it I think. I've made some more changes so I'll do the change. This is good work! I can merge this MR if you feel this is in a good state.

**Pull request merged**